### PR TITLE
Bring ukey2 up-to-date with internal code

### PR DESCRIPTION
**Pull request by apolyudov**
Signed-off-by: Alexey Polyudov <apolyudov@google.com>
Change-Id: Ie087ddaeef2c38f413484145114da7835c19f69b

### CODE REVIEW

File: .gitmodules [hunk 1]
```diff
@@ -0,0 +1,12 @@
+[submodule "third_party/secure_message"]
+	path = third_party/secure_message
+	url = https://github.com/apolyudov/securemessage
```
[the remark this answers is not in the document]
Comment (apolyudov) [reply .gitmodules]: 1. done.
2. done.
NOTE:
while doing the (2) I realized there is not java tests in this project, so independent verification of (2) is not possible. To resolve that, I added most lava tests to this project. (the only one I did not add was not added for the reason it had google3 .proto dependencies from different project).
Now, gradle build runs in siolation, and verification of both (1) and (2) is possible for non-googlers.

[the remark this answers is not in the document]
Comment (apolyudov) [reply .gitmodules]: Since changes are massive (I refreshed java code and added java tests), I published them as a separate commit  in the same PR.

File: gradlew [hunk 2]
```diff
@@ -0,0 +1,183 @@
+#!/usr/bin/env sh
```
Comment (leshi) [inline gradlew:1]: do we need all these gradle files?  can't they be pulled in via some submodule?

File: src/main/java/com/google/security/cryptauth/lib/securegcm/D2DConnectionContext.java [hunk 3]
```diff
@@ -1,17 +1,17 @@
-/* Copyright 2018 Google LLC
- *
```
Comment (leshi) [inline src/main/java/com/google/security/cryptauth/lib/securegcm/D2DConnectionContext.java:2]: It would be really nice if it were possible to make the java changes a separate PR.

File: src/main/java/com/google/security/cryptauth/lib/securegcm/D2DSpakeEd25519Handshake.java [hunk 4]
```diff
@@ -32,64 +32,59 @@
...
+ *
+ *
+ *
+ *
+ *
+ *
+ *
+ *
+ *
+ *  DO NOT USE, THIS CODE IS INSECURE (see b/138593608).
```
Comment (leshi) [inline src/main/java/com/google/security/cryptauth/lib/securegcm/D2DSpakeEd25519Handshake.java:61]: please delete this file and purge the commit history wrt this file

File: src/main/java/com/google/security/cryptauth/lib/securegcm/Ed25519.java [hunk 5]
```diff
@@ -208,20 +207,20 @@ public static void validateAffinePoint(BigInteger[] p) throws Ed25519Exception {
     };
   }
 
-  /**
-   * Converts a point in affine representation to extended representation
-   */
-  @VisibleForTesting
+  /** Converts a point in affine representation to extended representation */
+  // TODO(b/120887495): This @VisibleForTesting annotation was being ignored by prod code.
```
Comment (leshi) [inline src/main/java/com/google/security/cryptauth/lib/securegcm/Ed25519.java:211]: please search-replace references to internal bug numbers

File: src/main/java/com/google/security/cryptauth/lib/securemessage/CryptoOps.java [hunk 6]
```diff
@@ -491,27 +511,54 @@ static SecretKey deriveAes256KeyFor(SecretKey masterKey, String purpose)
...
     } catch (InvalidKeyException e) {
       throw new AssertionError(e);  // This should never happen
     }
-    // Arbitrary "info" to be included in the MAC.
-    macScheme.update(info);
-    // Note that RFC 5869 computes number of blocks N = ceil(hash length / output length), but
-    // here we only deal with a 256 bit hash up to a 256 bit output, yielding N=1.
-    return macScheme.doFinal(CONSTANT_01);
+
+    // Number of blocks N = ceil(hash length / output length).
```
Comment (leshi) [inline src/main/java/com/google/security/cryptauth/lib/securemessage/CryptoOps.java:531]: where are all these changes coming from?  Maybe the PR needs to be rebased?

File: gradlew (hunk 2 above)
Comment (apolyudov) [reply gradlew:1]: strictly speaking they are not necessary, but this a gradle bootstrapping code that makes life of a developer much easier. E.g. without this one I'm not able to build this project until and unless I somehow manage to create it (in my case, I'm using IntelliJ Idea to generate it for me; it also simplifies the readme steps: simply run the wrapper, otherwise it will include references to gradle docs and may be tricky for someone who is not developing Java code but wants to run the Java tests).

File: src/main/java/com/google/security/cryptauth/lib/securegcm/Ed25519.java (hunk 5 above)
Comment (apolyudov) [reply src/main/java/com/google/security/cryptauth/lib/securegcm/Ed25519.java:211]: if this is an OSS-related concern, it should not be, because cross tool says it is ok to have b/ references on OSS code. It helps tracking issues. If there is a different project-specific reason, I could scrub that.
